post: add post "Build an OCI Image from Scratch" (#26)
danishprakash/danishpraka.sh: pages/posts/build-oci-image-from-scratch.md

```
$ vim manifest.json
{
    "schemaVersion": 2,
    "mediaType": "application/vnd.oci.image.manifest.v1+json",
    "config": {
        "mediaType": "application/vnd.oci.image.config.v1+json",
        "digest": "sha256:bf7031d43f7d2c9ec77ed23de9d4daa3ed1482c4bd5eb4b941caee79f2c4be42",
        "size": 296
    },
    "layers": [
        {
            "mediaType": "application/vnd.oci.image.layer.v1.tar+gzip",
            "digest": "sha256:583eb9106f1be6dffa3f0e2fabe1fee5073878fbc66f7d2d5b19192db187e73a",
            "size": 1377804
        }
    ]
}

$ mv manifest.json $(sha256sum manifest.json | awk '{print $1}')
$ tree ../..
└── blobs/sha256
        ├── 38f01cd4419e646946527861333e01d71775c43965ff340dea0466eae61c200d
        ├── c37c06cdec9d6a0f2a2d55deb5aa002b26b37b17c02c2eca908fc062af5f53eb
        └── cd12bca58eb0ed29b24e4264179a2696b68df24f9006bc16563fe3525210936e
```

The `mediaType` attribute references the type of the media. OCI spec defines media types for various kinds of content. In our case, we refer to the config and layer with the relevant `mediaType` value. And as is required, we encode the manifest file too.

#### index.json

> Like how manifest.json acts as an index and metadata repository for an image for a specific architecture and operating system, an index.json file does that for the manifest.json files. In other words, it acts as an index for a set of images that can span different architectures and operating systems.

Reviewer: ```suggestion
index.json file acts as an index for a set of images that can span different architectures and operating systems.
```

The first sentence adds nothing of value, so you can drop it.

<img src="./../../static/img/posts/build-oci-image-from-scratch/index_multi.png"/>

> A simple example could be a container image that has different flavors e.g. a fully-featured base image with all the bells and whistles but also a "slim" variant that comes without unwanted additions. Here, the index.json would contain two entries pointing to the two variants of the image via their manifests, which would then contain image specific metadata.

Reviewer: ```suggestion
An example could be a container image that has different flavors: a fully featured base image and its minimal variant. In this case, the index.json would contain two entries pointing to the two variants of the image via their manifests that in turn would contain image specific metadata.
```

```
$ cd ../..
$ vim index.json
{
    "schemaVersion": 2,
    "manifests": [
        {
            "mediaType": "application/vnd.oci.image.manifest.v1+json",
            "digest": "sha256:38f01cd4419e646946527861333e01d71775c43965ff340dea0466eae61c200d",
            "size": 530,
            "annotations": {
                "org.opencontainers.image.ref.name": "hello:scratch"
            }
        }
    ]
}
```

> We've added an annotation representing the name and tag for our image referenced by the manifest. The index.json file isn't required to be encoded and therefore resides at the top-level directory of our image and not under `blobs/`. With all the finished components, a quick recap of what's what:

Reviewer: ```suggestion
We've added an annotation representing the name and tag for our image referenced by the manifest. The index.json file isn't required to be encoded, so it resides at the top-level directory of our image and not under `blobs/`. With all the finished components, let's do a quick recap:
```

```
$ tree
├── blobs/sha256
│       ├── 38f01cd4419e646946527861333e01d71775c43965ff340dea0466eae61c200d
│       ├── c37c06cdec9d6a0f2a2d55deb5aa002b26b37b17c02c2eca908fc062af5f53eb
│       └── cd12bca58eb0ed29b24e4264179a2696b68df24f9006bc16563fe3525210936e
└── index.json
```

(in the order shown above)

1. `38f01cd4419e646946527861333e01d71775c43965ff340dea0466eae61c200d` — Metadata such as image name and tag for our image along with other data.

> 2. `c37c06cdec9d6a0f2a2d55deb5aa002b26b37b17c02c2eca908fc062af5f53eb` — gzip compressed tar archive that consists of the only layer in our image which in turn consists of a single static C binary.

Reviewer: ```suggestion
2. `c37c06cdec9d6a0f2a2d55deb5aa002b26b37b17c02c2eca908fc062af5f53eb` — gzip compressed tar archive that consists of a single layer in the image. The layer consists of a single static C binary.
```

> 3. `cd12bca58eb0ed29b24e4264179a2696b68df24f9006bc16563fe3525210936e` — Container configuration options such as the entrypoint and environment variables that are going to be used by the container runtime when running the container based off of our image.

Reviewer: ```suggestion
3. `cd12bca58eb0ed29b24e4264179a2696b68df24f9006bc16563fe3525210936e` — Container configuration options, such as the entrypoint and environment variables, that are going to be used by the container runtime when running the container based on our image.
```

4. `index.json` — a higher-level manifest that references multiple image manifests, allowing for the distribution of container images across different platforms or architectures.

# Packing

> We have all the parts in place for our image. Let's create an archive of the various components we've created so far and test our image out:

Reviewer: ```suggestion
We have all the parts in place for our image. Let's create an archive of the components we've created so far, and then test our image:
```

```
$ tree
├── blobs/sha256
│       ├── 38f01cd4419e646946527861333e01d71775c43965ff340dea0466eae61c200d
│       ├── c37c06cdec9d6a0f2a2d55deb5aa002b26b37b17c02c2eca908fc062af5f53eb
│       └── cd12bca58eb0ed29b24e4264179a2696b68df24f9006bc16563fe3525210936e
└── index.json
$ tar -cf hello.tar *

$ podman load < hello.tar
Getting image source signatures
Copying blob c37c06cdec9d done   |
Copying config cd12bca58e done   |
Writing manifest to image destination
Loaded image: localhost/hello:scratch

$ podman run localhost/hello:scratch world
hello world!
```

> We can see that the image was loaded without any problems by podman, and we were able to run a container based on our image, correctly dumping `hello world!` to stdout. Let's inspect our loaded image:

Reviewer: ```suggestion
We can see that the image was loaded without any problems by podman, and we were able to run a container based on our image, correctly outputting `hello world!` to stdout. Let's inspect our loaded image:
```

```
$ podman image ls hello
REPOSITORY   TAG       IMAGE ID       CREATED   SIZE
hello        scratch    25e8b3bd9720  N/A       3.67MB
```

> Thanks to the scratch base image, our image has a minimal footprint, in fact, the only file in our image is our binary.

Reviewer: ```suggestion
Thanks to the scratch base image, our image has a minimal footprint. In fact, the only file in the image is our binary.
```

# Using a Base Image

> Let's try creating a version of our image which is based on `alpine` rather than scratch. This would introduce another layer in our image, so let's see how to handle multiple layers. For the sake of better understanding, this is how our Containerfile would look using alpine if we hadn't been doing it from scratch.

Reviewer: ```suggestion
Let's create a version of our image based on `alpine` instead of scratch. This introduces another layer in our image, so let's see how to handle multiple layers. In this case, the Containerfile is as follows.
```

```
FROM alpine:latest

COPY ./hello /root/hello

ENTRYPOINT ["time", "./hello"]
```

> Note that we also changed the entrypoint to use the `time` utility, part of the alpine builtins, this is simply to demonstrate working of the two layers in the image.

Reviewer: ```suggestion
Note that we changed the entrypoint to use the `time` utility. This is done to demonstrate how two layers work in the image.
```

> In our image so far, we have had just one layer that consists of the `hello` binary. As we've seen before, a layer can be considered a filesystem diff. So if we plan to use alpine as the base image, we need the respective root filesystem onto which we'll base the second layer, the one with the binary. You can obtain the alpine root filesystem from their official website.

Reviewer: ```suggestion
So far, the image has only one layer that consists of the `hello` binary. As mentioned previously, a layer can be considered a filesystem diff. So if we use alpine as the base image, we need the respective root filesystem onto which we'll base the second layer (that is, the layer with the binary). You can obtain the alpine root filesystem from the project's official website.
```

```
$ cd blobs/sha256
$ wget https://dl-cdn.alpinelinux.org/alpine/v3.18/releases/x86_64/alpine-minirootfs-3.18.4-x86_64.tar.gz
$ mv alpine-minirootfs-3.18.4-x86_64.tar.gz \
    $(sha256sum alpine-minirootfs-3.18.4-x86_64.tar.gz | awk '{print $1}')
```

That rootfs downloaded is already gzipped, so all we need to do is encode this with its sha256 digest and it's ready to go.

```
$ cd ../..
$ tree
├── blobs/sha256
│       ├── 75b148a9a5b61403d082f1930ccc779bfdac5acddb48ed8f90ccdc4219d51268
│       ├── 8289bd1bdc2a1fae2d2d717b7c40baaedc4c5c4d9c9f4f1a1b045287067e9f2c
│       ├── c37c06cdec9d6a0f2a2d55deb5aa002b26b37b17c02c2eca908fc062af5f53eb
│       └── c59d5203bc6b8b6ef81f3f6b63e32c28d6e47be806ba8528f8766a4ca506c7ba
└── index.json
```

> It's a little hard to figure out what's what with all those digests, but rest assured, you will _never_ find yourself doing this manually, ever again.

Reviewer: I'm not sure why you need this sentence. If you promise that you don't have to do that manually ever again, then explain why. I'd just drop this sentence.

> Now, on to updating our config and manifest files with the new layer. Once done, we also need to make sure, we encode them again with their respective updated sha256 digests.

Reviewer: ```suggestion
Next, we need to update config and manifest files with the new layer. After that, we also need to make sure that we encode them again with their respective updated sha256 digests.
```

```
$ vim config.json
{
    "architecture": "amd64",
    "os": "linux",
    "config": {
        "Entrypoint": [
            "time",
            "./hello"
        ]
    },
    "rootfs": {
        "type": "layers",
        "diff_ids": [
--------->  "sha256:c59d5203bc6b8b6ef81f3f6b63e32c28d6e47be806ba8528f8766a4ca506c7ba",
            "sha256:c37c06cdec9d6a0f2a2d55deb5aa002b26b37b17c02c2eca908fc062af5f53eb"
        ]
    }
}

$ mv config.json $(sha256sum config.json | awk '{print $1}')
```

> Layers are represented top to bottom so the base alpine rootfs layer comes first and then the second layer with the binary. Note also that we modified the entrypoint to use the `time` utility as discussed previously. The manifest file also sees similar changes--adding the layer digests--along with updating the size of our config file, and not to forget re-calculating the digest and encoding our updated manifest.

Reviewer: ```suggestion
Layers are represented top to bottom so the base alpine rootfs layer comes first, and then the second layer with the binary. Keep in mind that we modified the entrypoint to use the `time` utility. The manifest file also sees similar changes--adding the layer digests--along with updating the size of our config file, and not to forget re-calculating the digest and encoding our updated manifest.
```

The last sentence is kind of clunky, but I'm not sure how to rewrite it. Maybe _the manifest contains the following changes: 1) adds layer digest, 2) ..._?

```
$ vim manifest.json
{
    "schemaVersion": 2,
    "mediaType": "application/vnd.oci.image.manifest.v1+json",
    "config": {
        "mediaType": "application/vnd.oci.image.config.v1+json",
----->  "digest": "sha256:75b148a9a5b61403d082f1930ccc779bfdac5acddb48ed8f90ccdc4219d51268",
----->  "size": 403
    },
    "layers": [
        {
            "mediaType": "application/vnd.oci.image.layer.v1.tar+gzip",
            "digest": "sha256:c59d5203bc6b8b6ef81f3f6b63e32c28d6e47be806ba8528f8766a4ca506c7ba",
            "size": 3279768
        },
        {
            "mediaType": "application/vnd.oci.image.layer.v1.tar+gzip",
--------->  "digest": "sha256:c37c06cdec9d6a0f2a2d55deb5aa002b26b37b17c02c2eca908fc062af5f53eb",
            "size": 1372611
        }
    ]
}

$ mv manifest.json $(sha256sum manifest.json | awk '{print $1}')
```

> Note: the filenames for both config.json and manifest.json would've been the sha256 digests from the first variant of the image we built, I'm choosing to revert them to config.json and manifest.json for presentation purposes.

Reviewer: _would have been_ means that they are actually not. For example: It would have been if you come (but you didn't, so it wasn't nice). Do you mean _should be_?

Author: You're right, fixed!

--------->  "size": 748,
            "annotations": {
                "org.opencontainers.image.ref.name": "hello:alpine1"
            }
        }
    ]
}
```

If we pack and load the image again with podman, we can see the modified image in action:
```
$ tree
├── blobs
│   └── sha256
│       ├── 75b148a9a5b61403d082f1930ccc779bfdac5acddb48ed8f90ccdc4219d51268
│       ├── 8289bd1bdc2a1fae2d2d717b7c40baaedc4c5c4d9c9f4f1a1b045287067e9f2c
│       ├── c37c06cdec9d6a0f2a2d55deb5aa002b26b37b17c02c2eca908fc062af5f53eb
│       └── c59d5203bc6b8b6ef81f3f6b63e32c28d6e47be806ba8528f8766a4ca506c7ba
└── index.json
$ tar -cvf hello-alpine.tar *

$ podman load < hello-alpine.tar
Getting image source signatures
Copying blob c59d5203bc6b done   |
Copying blob c37c06cdec9d done   |
Copying config 75b148a9a5 done   |
Writing manifest to image destination
Loaded image: localhost/hello:alpine

$ podman run --rm localhost/hello:alpine reader
Hello, reader!
real    0m 0.00s
user    0m 0.00s
sys     0m 0.00s

$ podman image ls hello
REPOSITORY   TAG       IMAGE ID       CREATED   SIZE
hello        scratch   25e8b3bd9720   N/A       3.67MB
hello        alpine    3d0268e9a91e   N/A       11MB
```

> So there we have it, our two images loaded into the docker daemon. We can run both of them and see the same output from both of them. One interesting thing to note is the size difference between the two images. Alpine, while still having a lower footprint is larger than scratch, but nothing extraordinary. Hopefully this example demonstrated interaction between different layers in an image and how a manifest file can represent multiple images from a single file.

Reviewer: > One interesting thing to note is the size difference between the two images. Alpine, while still having a lower footprint is larger than scratch, but nothing extraordinary.

Why is it interesting? And why do you have to mention it at all?

> Hopefully this example demonstrated interaction between different layers in an image and how a manifest file can represent multiple images from a single file.

Drop this sentence. If you have to explain in the end what you were trying to achieve, you failed to achieve it. :smile:

# Conclusion
I was initially planning to title this article "Container Image Internals" and the first draft of that came out all theoretical and while reading it, even I felt bored even though I enjoyed finding about all I could about container images for the post. But then I decided to document the fun part of this whole journey and here we are. In this article, we saw how we can "build" a docker image from scratch. But that's not to say that we'd be doing something like that on a day to day basis. You should never find yourself doing this as part of your regular development workflow. This article intentionally uses a simple example and so it misses out on important nuances such as how layers use an impelementation of UnionFS and don't pack simplified tarballs anymore, etc.

> But alas, it's always fun to crack open something and to understanding things so I hope this article helped demistify container images to a certain extent. If you find anything that seems wrong or if you think there's something that can be improved, feel free to reach out.

Reviewer: Keep it tight and simple.

1. The examples used in this articles is not something you'd do in your daily work. Their purpose is to demonstrate of inner working of [insert what inner workings].
2. Reach out.
3. Done!
